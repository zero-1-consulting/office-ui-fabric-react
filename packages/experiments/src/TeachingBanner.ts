export * from './components/TeachingBanner/index';
