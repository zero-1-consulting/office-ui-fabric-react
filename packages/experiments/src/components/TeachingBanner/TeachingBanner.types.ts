import { BaseButton, Button, IStackSlot, IButtonProps, ISchemeNames, ITextSlot } from 'office-ui-fabric-react';
import { IComponent, IComponentStyles, IStyleableComponentProps, ISlottableProps, ISlotProp, IHTMLSlot } from '@uifabric/foundation';
import { IFontIconSlot } from '../../utilities/factoryComponents.types';

/**
 * {@docCategory TeachingBanner}
 */
export type ITeachingBannerComponent = IComponent<
  ITeachingBannerProps,
  ITeachingBannerTokens,
  ITeachingBannerStyles,
  ITeachingBannerViewProps
>;

// The following two types are redundant with ITeachingBannerComponent but are needed until TS function return widening issue is resolved:
// https://github.com/Microsoft/TypeScript/issues/241
// For now, these helper types can be used to provide return type safety for tokens and styles functions.

/**
 * {@docCategory TeachingBanner}
 */
export type ITeachingBannerTokenReturnType = ReturnType<Extract<ITeachingBannerComponent['tokens'], Function>>;

/**
 * {@docCategory TeachingBanner}
 */
export type ITeachingBannerStylesReturnType = ReturnType<Extract<ITeachingBannerComponent['styles'], Function>>;

/*
 * Optional interface to use for componentRef. This should be limited in scope with the most common scenario being for focusing elements.
 * {@docCategory TeachingBanner}
 */
export interface ITeachingBanner {}

/**
 * {@docCategory TeachingBanner}
 */
export interface ITeachingBannerSlots {
  /**
   * Root element.
   */
  root?: IHTMLSlot;

  /**
   * Defines the icon that is displayed in Premium mode.
   * @defaultValue defaultPremiumIcon
   */
  iconPremium?: IFontIconSlot;

  /**
   * Text to add as a headline.
   */
  headline?: ITextSlot;

  /**
   * Content for Teaching Banner
   */
  content?: IStackSlot;

  /**
   * Container for action buttons.
   */
  actionsContainer?: IStackSlot;

  /**
   * Action PrimaryButton Slot.
   */
  actionPrimaryButton?: ISlotProp<IButtonProps>;

  /**
   * Action DefaultButton Slot.
   */
  actionDefaultButton?: ISlotProp<IButtonProps>;

  /**
   * Dismiss IconButton Slot.
   */
  dismissButton?: ISlotProp<IButtonProps>;
}

/**
 * Extending IStyleableComponentProps will automatically add styleable props for you, such as styles, tokens and theme.
 * If you don't want these props to be included in your component, just remove this extension.
 * {@docCategory TeachingBanner}
 */
export interface ITeachingBannerProps
  extends ISlottableProps<ITeachingBannerSlots>,
    IStyleableComponentProps<ITeachingBannerViewProps, ITeachingBannerTokens, ITeachingBannerStyles> {
  /**
   * Changes scheme for sub components
   * @defaultValue strong
   */
  scheme?: ISchemeNames;

  /**
   * Component default Premium icon for use if component is uncontrolled.
   * @defaultValue Diamond
   */
  defaultPremiumIcon?: string;

  /**
   * The action buttons' props you want to show on the other side.
   */
  actions?: IButtonProps[];

  /**
   * Set Premium mode. Shows the banner's icon and thickness is bigger.
   * @defaultValue false
   */
  premium?: boolean;

  /**
   * Enables wrapping text and action buttons.
   * @defaultValue true
   */
  multiline?: boolean;

  /**
   * Whether the Teaching Banner has a dismiss button and its callback.
   * If null, we don't show a dismiss button.
   * @defaultvalue null
   */
  onDismiss?: (ev?: React.MouseEvent<HTMLElement | BaseButton | Button>) => any;
}

/**
 * {@docCategory TeachingBanner}
 */
export interface ITeachingBannerViewProps extends ITeachingBannerProps {
  /**
   * Sample props internal to component. These types of props aren't exposed
   * externally to consumers and their values are typically determined by component state.
   */
  dismissed?: boolean;
}

/**
 * {@docCategory TeachingBanner}
 */
export interface ITeachingBannerTokens {
  /**
   * Token for Premium icon, text and dismissButton color.
   */
  color?: string;

  /**
   * Token for the root background;
   */
  background?: string;

  /**
   * Token for the spacing between elements;
   */
  gap?: number;
}

/**
 * {@docCategory TeachingBanner}
 */
export type ITeachingBannerStyles = IComponentStyles<ITeachingBannerSlots>;
