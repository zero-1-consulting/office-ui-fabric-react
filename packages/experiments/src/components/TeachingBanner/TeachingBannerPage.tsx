import * as React from 'react';
import { ExampleCard, IComponentDemoPageProps, ComponentPage, PropertiesTableSet } from '@uifabric/example-app-base';

import { TeachingBannerBasicExample } from './examples/TeachingBanner.Basic.Example';
const TeachingBannerBasicExampleCode = require('!raw-loader!@uifabric/experiments/src/components/TeachingBanner/examples/TeachingBanner.Basic.Example.tsx') as string;

export class TeachingBannerPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title="TeachingBanner"
        componentName="TeachingBanner"
        exampleCards={
          <div>
            <ExampleCard title="Folder TeachingBanner" isOptIn={true} code={TeachingBannerBasicExampleCode}>
              <TeachingBannerBasicExample />
            </ExampleCard>
          </div>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={[require<string>('!raw-loader!@uifabric/experiments/src/components/TeachingBanner/TeachingBanner.types.ts')]}
          />
        }
        overview={
          <div>
            <p>
              Teaching banners are used to educate users about a product. Unlike callouts, banners are used in cases when you canâ€™t point to
              a feature but instead need to educate on a concept or a set of features. These banners are dismissible and can contain call to
              action buttons.
            </p>
          </div>
        }
        bestPractices={<div />}
        dos={
          <div>
            <ul>
              <li />
            </ul>
          </div>
        }
        donts={
          <div>
            <ul>
              <li>Don't use to display errors, warning or info messages.</li>
            </ul>
          </div>
        }
        isHeaderVisible={this.props.isHeaderVisible}
      />
    );
  }
}
