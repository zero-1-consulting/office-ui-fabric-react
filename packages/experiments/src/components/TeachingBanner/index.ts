export * from './TeachingBanner';
export * from './TeachingBanner.types';
