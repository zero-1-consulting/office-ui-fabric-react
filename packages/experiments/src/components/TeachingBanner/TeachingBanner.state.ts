import { useCallback, useState } from 'react';
import { ITeachingBannerComponent, ITeachingBannerProps, ITeachingBannerViewProps } from './TeachingBanner.types';
import { useControlledState } from '../../Foundation';

/**
 * {@docCategory TeachingBanner}
 */
export const useTeachingBannerState: ITeachingBannerComponent['state'] = (
  props: Readonly<ITeachingBannerProps>
): ITeachingBannerViewProps => {
  const [dismissed, setDismissed] = useState(false);
  const [iconPremium] = useControlledState(props, 'iconPremium', {
    defaultPropName: 'defaultPremiumIcon',
    defaultPropValue: 'Diamond'
  });

  const { onDismiss } = props;

  const _onDismiss = useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      setDismissed(true);
      onDismiss!(event);
    },
    [dismissed]
  );

  const viewProps: ITeachingBannerViewProps = {
    ...props,
    iconPremium,
    dismissed,
    onDismiss: (onDismiss && _onDismiss) || undefined
  };

  return viewProps;
};
