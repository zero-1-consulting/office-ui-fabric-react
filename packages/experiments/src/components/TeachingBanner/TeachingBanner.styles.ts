import { ITeachingBannerComponent, ITeachingBannerStylesReturnType, ITeachingBannerTokenReturnType } from './TeachingBanner.types';
import { getGlobalClassNames } from '../../Styling';

const GlobalClassNames = {
  root: 'ms-TeachingBanner',
  icon: 'ms-TeachingBanner-icon',
  headline: 'ms-TeachingBanner-headline',
  content: 'ms-TeachingBanner-content',
  actions: 'ms-TeachingBanner-actions',
  dismiss: 'ms-TeachingBanner-dismiss'
};

// const singleLineProps: IRawStyle = { overflowX: 'hidden', whiteSpace: 'nowrap' };

/**
 * {@docCategory TeachingBanner}
 */
export const TeachingBannerTokens: ITeachingBannerComponent['tokens'] = (props, theme): ITeachingBannerTokenReturnType => [];

/**
 * {@docCategory TeachingBanner}
 */
export const TeachingBannerStyles: ITeachingBannerComponent['styles'] = (props, theme, tokens): ITeachingBannerStylesReturnType => {
  const classNames = getGlobalClassNames(GlobalClassNames, theme);
  const { palette } = theme;
  const { onDismiss, multiline = false } = props;
  const { gap = 20 } = tokens;

  return {
    root: [
      classNames.root,
      {
        position: 'relative',
        display: 'flex',
        alignItems: 'center',
        paddingRight: (onDismiss && !multiline && 55) || 14,
        paddingLeft: 14,
        width: '100%',
        minHeight: ((props.premium || props.actions) && 60) || 40,
        borderRadius: 2,
        color: tokens.color || palette.white,
        background: tokens.background || palette.themePrimary,
        boxSizing: 'border-box'
      }
    ],
    content: [
      classNames.content,
      {
        flex: '1',
        display: 'flex',
        selectors: {
          '&>*:not(.ms-TeachingBanner-icon):not(.ms-TeachingBanner-heading)': {
            overflowX: 'hidden',
            textOverflow: 'ellipsis'
          }
        }
      },
      (multiline && {
        flexWrap: 'wrap',
        marginTop: 25,
        flexDirection: 'column',
        alignItems: 'flex-start'
      }) || { flexWrap: 'nowrap', alignItems: 'center', overflowX: 'hidden', whiteSpace: 'nowrap', textOverflow: 'ellipsis' }
    ],
    iconPremium: [
      classNames.icon,
      {
        flex: '0 0 auto',
        fontSize: 24,
        marginRight: 17,
        marginLeft: 6
      },
      multiline && { alignSelf: 'flex-start', marginTop: 22 }
    ],
    headline: [classNames.headline, { flex: '0 0 auto', marginRight: 16, fontWeight: 600 }],
    actionsContainer: [
      classNames.actions,
      {
        flex: '0 0 auto',
        marginLeft: 'auto',
        selectors: {
          '.ms-Button': {
            marginLeft: gap,
            marginTop: 14,
            marginBottom: 14
          }
        }
      }
    ],
    dismissButton: [
      classNames.dismiss,
      {
        position: 'absolute',
        right: 0,
        top: 0,
        margin: 8,
        width: 24,
        height: 24,
        selectors: {
          '.ms-Button-icon': {
            fontSize: 12,
            height: 12,
            lineHeight: 12
          },
          ':hover': {
            background: 'transparent'
          },
          ':active': {
            background: 'transparent'
          }
        }
      }
    ]
  };
};
