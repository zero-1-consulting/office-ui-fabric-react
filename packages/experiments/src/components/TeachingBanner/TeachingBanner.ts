import { TeachingBannerView } from './TeachingBanner.view';
import { TeachingBannerStyles, TeachingBannerTokens } from './TeachingBanner.styles';
import { useTeachingBannerState } from './TeachingBanner.state';
import { ITeachingBannerProps } from './TeachingBanner.types';
import { createComponent } from '../../Foundation';

/**
 * {@docCategory TeachingBanner}
 */
export const TeachingBanner: React.StatelessComponent<ITeachingBannerProps> = createComponent(TeachingBannerView, {
  displayName: 'TeachingBanner',
  state: useTeachingBannerState,
  styles: TeachingBannerStyles,
  tokens: TeachingBannerTokens
});
