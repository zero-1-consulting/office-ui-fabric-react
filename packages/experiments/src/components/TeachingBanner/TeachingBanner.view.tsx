/** @jsx withSlots */
import { withSlots, getSlots, ThemeProvider } from '../../Foundation';
import { ITeachingBannerComponent, ITeachingBannerProps, ITeachingBannerSlots } from './TeachingBanner.types';
import { FontIcon } from '../../utilities/factoryComponents';
import { Stack, DefaultButton, PrimaryButton, IconButton, Text } from 'office-ui-fabric-react';

/**
 * {@docCategory TeachingBanner}
 */
export const TeachingBannerView: ITeachingBannerComponent['view'] = props => {
  const Slots = getSlots<ITeachingBannerProps, ITeachingBannerSlots>(props, {
    root: 'div',
    iconPremium: FontIcon,
    content: 'div',
    headline: Text,
    actionsContainer: Stack,
    actionPrimaryButton: PrimaryButton,
    actionDefaultButton: DefaultButton,
    dismissButton: IconButton
  });

  if (props.dismissed) {
    return null;
  }
  const { actions, children, premium, headline, onDismiss, scheme = 'strong' } = props;

  return (
    <ThemeProvider scheme={scheme}>
      <Slots.root>
        {premium && <Slots.iconPremium />}
        <Slots.content>
          {headline && <Slots.headline as="strong" />}
          {children}
          {actions && (
            <Slots.actionsContainer horizontal>
              {actions.map(itemProps =>
                itemProps.primary ? <Slots.actionPrimaryButton {...itemProps} /> : <Slots.actionDefaultButton {...itemProps} />
              )}
            </Slots.actionsContainer>
          )}
        </Slots.content>
        {onDismiss && <Slots.dismissButton iconProps={{ iconName: 'Cancel' }} onClick={onDismiss} />}
      </Slots.root>
    </ThemeProvider>
  );
};
