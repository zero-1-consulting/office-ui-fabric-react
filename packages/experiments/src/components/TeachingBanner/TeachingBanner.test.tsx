import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { TeachingBanner } from './TeachingBanner';
import { mount } from 'enzyme';

const noop = () => null;

describe('TeachingBanner', () => {
  it('renders correctly with no props', () => {
    const tree = renderer.create(<TeachingBanner />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders children prop correctly', () => {
    const tree = renderer.create(<TeachingBanner>Text</TeachingBanner>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders children prop correctly', () => {
    const component = mount(<TeachingBanner>Text</TeachingBanner>);
    expect(
      component
        .find('.ms-TeachingBanner-content')
        .first()
        .text()
    ).toEqual('Text');
  });

  it('matches children prop snapshot', () => {
    const tree = renderer.create(<TeachingBanner>Text</TeachingBanner>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render dismiss button', () => {
    const component = mount(<TeachingBanner onDismiss={noop} />);
    expect(component.find('.ms-TeachingBanner-dismiss').first()).toBeTruthy();
  });
});
