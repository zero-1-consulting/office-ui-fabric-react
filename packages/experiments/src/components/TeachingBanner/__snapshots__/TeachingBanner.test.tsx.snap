// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingBanner matches children prop snapshot 1`] = `
<div
  className=
      ms-TeachingBanner
      {
        align-items: center;
        background: #0078d4;
        border-radius: 2px;
        box-sizing: border-box;
        color: #ffffff;
        display: flex;
        min-height: 40px;
        padding-left: 14px;
        padding-right: 14px;
        position: relative;
        width: 100%;
      }
>
  <div
    className=
        ms-TeachingBanner-content
        {
          align-items: center;
          display: flex;
          flex-wrap: nowrap;
          flex: 1;
          overflow-x: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        &>*:not(.ms-TeachingBanner-icon):not(.ms-TeachingBanner-heading) {
          overflow-x: hidden;
          text-overflow: ellipsis;
        }
  >
    Text
  </div>
</div>
`;

exports[`TeachingBanner renders children prop correctly 1`] = `
<div
  className=
      ms-TeachingBanner
      {
        align-items: center;
        background: #0078d4;
        border-radius: 2px;
        box-sizing: border-box;
        color: #ffffff;
        display: flex;
        min-height: 40px;
        padding-left: 14px;
        padding-right: 14px;
        position: relative;
        width: 100%;
      }
>
  <div
    className=
        ms-TeachingBanner-content
        {
          align-items: center;
          display: flex;
          flex-wrap: nowrap;
          flex: 1;
          overflow-x: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        &>*:not(.ms-TeachingBanner-icon):not(.ms-TeachingBanner-heading) {
          overflow-x: hidden;
          text-overflow: ellipsis;
        }
  >
    Text
  </div>
</div>
`;

exports[`TeachingBanner renders correctly with no props 1`] = `
<div
  className=
      ms-TeachingBanner
      {
        align-items: center;
        background: #0078d4;
        border-radius: 2px;
        box-sizing: border-box;
        color: #ffffff;
        display: flex;
        min-height: 40px;
        padding-left: 14px;
        padding-right: 14px;
        position: relative;
        width: 100%;
      }
>
  <div
    className=
        ms-TeachingBanner-content
        {
          align-items: center;
          display: flex;
          flex-wrap: nowrap;
          flex: 1;
          overflow-x: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        &>*:not(.ms-TeachingBanner-icon):not(.ms-TeachingBanner-heading) {
          overflow-x: hidden;
          text-overflow: ellipsis;
        }
  />
</div>
`;
